llvm/lib/Target/AMDGPU/AMDGPUCombine.td

```diff
@@ -33,6 +33,14 @@ def rcp_sqrt_to_rsq : GICombineRule<
          [{ return matchRcpSqrtToRsq(*${rcp}, ${matchinfo}); }]),
   (apply [{ Helper.applyBuildFn(*${rcp}, ${matchinfo}); }])>;
 
+def fdiv_1_by_sqrt_to_rsq : GICombineRule<
+  (defs root:$root),
+  (match (G_FSQRT $sqrt, $x, (MIFlags FmContract)),
+         (G_FCONSTANT $one, $fpimm),
+         (G_FDIV $dst, $one, $sqrt, (MIFlags FmContract)):$root,
+         [{ return ${fpimm}.getFPImm()->isExactlyValue(1.0)
+            || ${fpimm}.getFPImm()->isExactlyValue(-1.0); }]),
```

Comment:
How it is correct to match -1.0 here?

Reply:
I am a novice, so I am asking in hopes of learning, but what would make this incorrect? This seems to work as far as matching MIR that contains -1/sqrt(x) pattern.

Reply:
Converting 1/sqrt(x) to rsq(x) seems fine since that is the definition of rsq. Converting -1/sqrt(x) to rsq(x) seems wrong because that's just not what rsq does.

Reply:
Right, that makes sense. Could we do -1/sqrt(x) to fneg(rsq(x))?

Reply:
Yes, we do try to fold that in other places. AMDGPUCodeGenPrepare handles most of these folds, it just leaves the fastest math paths for codegen.

Reply:
> Could we do -1/sqrt(x) to fneg(rsq(x))?

Yes, especially since fneg is cheap since it can be folded into lots of other (AMDGPU) instructions. But it does make me wonder again, if you're going to generalise, why not generalise to any value? I.e. y/sqrt(x) -> y*rsq(x)

Reply:
> > Could we do -1/sqrt(x) to fneg(rsq(x))?
> 
> Yes, especially since fneg is cheap since it can be folded into lots of other (AMDGPU) instructions. But it does make me wonder again, if you're going to generalise, why not generalise to any value? I.e. y/sqrt(x) -> y*rsq(x)

I agree generalizing makes sense if you're going to handle more than 1/sqrt - you can simplify the apply function at least to just emit `y * rsq(x)` and let other combine eliminate a `1*rsq`, or transform `-1*rsq` into `fneg(rsq)`

Reply:
> I agree generalizing makes sense if you're going to handle more than 1/sqrt - you can simplify the apply function at least to just emit `y * rsq(x)` and let other combine eliminate a `1*rsq`, or transform `-1*rsq` into `fneg(rsq)`

So should I re-implement this PR to match `y/sqrt(x)` and have the apply function emit `y * rsq(x)`  and a second combine match `1 * rsq` & `-1 * rsq` to `rsq` & `fneg(rsq)` respectively?

Reply:
`right_identity_one_fp` will take care of the `1 * rsq`, it seems like we don't have one for `-1.0` to `G_FNEG` though so you can do it in a follow up patch if you want. `X * 1` and `X * -1` are generic simplifications that work for anything in X

So no, you don't need to reimplement much, just make the matcher more generic (don't check for +-1.0) and in the apply, you always emit `rsq(x) * y`. Prefer to emit the `y` on the rhs, that way if it's a constant it's already the canonical form for mul (constants are always on the RHS for commutative ops IIRC)

Reply:
Yeah I would like to add fold for `-1.0` to `G_FNEG` in a follow up patch.